facenet_sandberg/generate_pairs.py

```diff
@@ -4,139 +4,124 @@
 
 import os
 import random
-from typing import List, Tuple
-
+import sys
+import argparse
 import numpy as np
+from typing import List, Tuple, cast
+
 
-names = [
-    d for d in os.listdir(image_dir) if os.path.isdir(
-        os.path.join(image_dir, d))]
+Mismatch = Tuple[str, int, str, int]
+Match = Tuple[str, int, int]
 
 
-def split_people_into_sets(image_dir: str, k_num_sets: int) -> List[List[str]]:
-    names = [
-        d for d in os.listdir(image_dir) if os.path.isdir(
-            os.path.join(
-                image_dir,
-                d))]
+def write_pairs(fname: str,
+                match_folds: List[List[Match]],
+                mismatch_folds: List[List[Mismatch]],
+                k_num_sets: int,
+                total_matches_mismatches: int) -> None:
+    file_contents = f'{k_num_sets}\t{total_matches_mismatches}\n'
+    for match_fold, mismatch_fold in zip(match_folds, mismatch_folds):
+        for match in match_fold:
+            file_contents += f'{match[0]}\t{match[1]}\t{match[2]}\n'
+        for mismatch in mismatch_fold:
+            file_contents += f'{mismatch[0]}\t{mismatch[1]}\t\
+{mismatch[2]}\t{mismatch[3]}\n'
+    with open(fname, 'w') as fpairs:
+        fpairs.write(file_contents)
+
+
+def _split_people_into_folds(image_dir: str,
+                             k_num_sets: int) -> List[List[str]]:
+    names = [d for d in os.listdir(image_dir)
+             if os.path.isdir(os.path.join(image_dir, d))]
     random.shuffle(names)
+    return [list(arr) for arr in np.array_split(names, k_num_sets)]
 
 
-def make_matches(image_dir: str,
-                 people: List[str],
-                 total_matches: int) -> List[Tuple[str,
-                                                   int,
-                                                   int]]:
-    matches: List[Tuple[str, int, int]] = []
+def _make_matches(image_dir: str,
+                  people: List[str],
+                  total_matches: int) -> List[Match]:
+    matches = cast(List[Match], [])
```

Comment:
Given that you're already assuming Python 3.6 (via the f strings), what does the cast buy us over a variable annotation? `matches: List[Match] = []`

Reply:
In the meeting today, we talked about trying to exclusively use type hinting for function arguments. When mypy complains, we decided to use cast instead of using types in the functions. I am happy to reopen that discussion/switch to not using cast, lmk which you prefer.

Reply:
In general, avoid variable annotations. If the type checker complains, do whatever is the least hassle to get it to shut up (in this case that'd be a variable annotation or type comment). The `cast` is just odd -- types are a pre-runtime concern, so why make them part of the runtime?